### ISSUE TITLE
NaNs in Gradient of RGB to LAB Transformation

### ISSUE THREAD

**Issue by zacharyjgs** [kornia/kornia#868]
## To Reproduce

1. Apply `rgb_to_lab` transform to image.
2. Compute gradient of sum over new image pixels with respect to original input image.

```
x_new = kornia.color.rgb_to_lab(x)
grad = torch.autograd.grad(x_new.sum(), x)[0].detach()
```

## Expected behavior

There were no NaNs in the output image `x_new`, so if this transform is differentiable, there should be no NaNs in the gradient.

## Environment

PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 450.57
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] pytorch-colors==0.1.0
[pip3] torch==1.7.1
[pip3] torchvision==0.8.1
[conda] Could not collect

**Comment by edgarriba** [comment]
@zacharyjgs thanks for pointing this out. /cc @cceyda Could you take a quick look at this ?

I was checking, and this [line](https://github.com/kornia/kornia/blob/master/kornia/color/lab.py#L56) in the case everything is zero might produce that. Not sure though - @zacharyjgs what is your input ?

**Comment by cceyda** [comment]
@zacharyjgs could you check if your input for `rgb_to_lab` has values not in range [0,1]

**Comment by zacharyjgs** [comment]
Thanks so much! Yes, I definitely have zeros in my input so this is the issue!

**Comment by zacharyjgs** [comment]
Also @cceyda, I was also wondering if the line `torch.pow(xyz_normalized, 1 / 3)` should be `torch.pow(xyz_normalized, 1 / 3.)`? Since, `1 / 3` will evaluate to `0`, right?

**Comment by edgarriba** [comment]
@cceyda thanks for pointing this out. I mark this as a pending issue to solve in next releases.

**Comment by manipopopo** [comment]
Maybe we could change 
https://github.com/kornia/kornia/blob/cf8e854546110e3dac1caa35449429fc7af328e5/kornia/color/lab.py#L58-L60
to
```python
threshold = 0.008856
xyz_int = torch.where(
    xyz_normalized > threshold,
    xyz_normalized.clamp_min(threshold) ** (1.0 / 3.0), 
    7.787 * xyz_normalized + 4. / 29.
)
```

And maybe the relevant
https://github.com/kornia/kornia/blob/cf8e854546110e3dac1caa35449429fc7af328e5/kornia/color/lab.py#L130-L134
could be changed to
```python
threshold = 0.0031308
rgb_im = torch.where(rgbs_im > threshold, 1.055 * torch.pow(rgbs_im.clamp_min(threshold), 1 / 2.4) - 0.055, 12.92 * rgbs_im )
```

**Comment by edgarriba** [comment]
@cceyda any final decision on this ?

**Comment by cceyda** [comment]
I think using `clamp_min` like @manipopopo suggests is the way to go, since the values that cause the nans never satisfy the where condition anyway.
I will also factor out the linear RGB to sRGB stuff (L130-134 ) so that `rgb_to_luv` is also fixed.
Expect a PR within today~

**Comment by sreenithy** [comment]
I added the `clamp`  function as suggested. Still get the Nan in gradients

**Comment by sreenithy** [comment]
@cceyda It was the range issue. Sorted thank you.
